### PR TITLE
GenQ infrastructure

### PR DESCRIPTION
**Pull request by yfguo**
## Pull Request Description

This PR adds a new pool/queue infrastructure, called GenQ.  The POSIX/iqueue usage is replaced by this new implementation.

There are four concepts introduced in this PR: shmem pool, shmem queue, private pool, and cells.

Pools have storage associated with them in the form of cells, and allow for cells to allocated and freed.  Cells in a pool are not ordered and can be allocated in arbitrary order.  Queues do not have any storage associated with them, but they are FIFO ordered.  A cell that is allocated from a pool can be enqueued into or dequeued from a queue.

Shmem pools are backed by shared memory storage.  Private pools are backed by user-defined memory (through a memory allocation function).  For both pools, the actual allocation of the storage is performed by the GenQ infrastructure.

|                       |  Multi-process safety | Multi-thread safety | Allocation | Pool Size |
|--------------|-----------------------|--------------------|-----------|------|
| Shmem pool |   MPSC                         |  No                            |  Collective | Fixed |
| Shmem queue | MPSC or SERIAL      |  No                            |  Local        | N.A.  |
| Private pool  |   N.A.                            |  No                            |  Local        | Dynamic |

The current version of the code does not provide any thread safety.

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] You or your company has a signed contributor's agreement on file with Argonne
* [ ] For non-Argonne authors, request an explicit comment from your companies PR approval manager

### CODE REVIEW

File: src/mpid/common/iqueue/mpidu_iqueue.c [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+    MPL_atomic_store_int(&new_pool->cellblk_mgr->next_free_idx, 0);
+
+    *pool = new_pool;
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_POOL_ALLOC);
+    return mpi_errno;
+  fn_fail:
+    if (new_pool) {
```
Comment (gcongiu) [inline src/mpid/common/iqueue/mpidu_iqueue.c:65]: C99 allows to pass `NULL` (7.20.3.2/2 from http://www.open-std.org/JTC1/SC22/wg14/www/docs/n1124.pdf) to free so these checks can be removed.

File: src/mpid/common/iqueue/mpidu_iqueue.c [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+    if (pool->slab) {
+        pool->free_fn(pool->slab);
+    }
+    MPL_free(pool);
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_POOL_FREE);
+}
+
+void MPIDU_iqueue_queue_alloc(MPIDU_iqueue_pool_t * pool, int index, MPIDU_iqueue_queue_t ** queue)
```
Comment (gcongiu) [inline src/mpid/common/iqueue/mpidu_iqueue.c:87]: This function actually does not allocate any memory just returns the pointer to the queue having requested index. Should this be renamed `MPIDU_iqueue_get_queue` or `MPIDU_iqueue_init_queue`?

File: src/mpid/common/iqueue/mpidu_iqueue.c [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    MPIDU_iqueue_queue_t *new_queue = pool->all_queues + sizeof(MPIDU_iqueue_queue_t) * index;
+    new_queue->u.handle = MPIDU_IQUEUE_POOL_GET_OBJ_HANDLE(pool, new_queue);
+    new_queue->u.head = NULL;
+    MPL_atomic_relaxed_store_ptr(&new_queue->u.tail, NULL);
+    *queue = new_queue;
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_QUEUE_ALLOC);
+}
+
+void MPIDU_iqueue_queue_free(MPIDU_iqueue_queue_t * queue)
```
Comment (gcongiu) [inline src/mpid/common/iqueue/mpidu_iqueue.c:103]: Same comment as above. What about `MPIDU_iqueue_put_queue` or `MPIDU_iqueue_fini_queue`?

File: src/mpid/common/iqueue/mpidu_iqueue.c [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+static void iqueue_cellblock_free(MPIDU_iqueue_pool_t * pool, MPIDU_iqueue_cell_t * cells)
+{
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_IQUEUE_CELLBLOCK_FREE);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_IQUEUE_CELLBLOCK_FREE);
+
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_IQUEUE_CELLBLOCK_FREE);
+    return;
+}
+
+void MPIDU_iqueue_cell_reserve(MPIDU_iqueue_pool_t * pool)
```
Comment (gcongiu) [inline src/mpid/common/iqueue/mpidu_iqueue.c:144]: This initializes all the cells in the pool so `cells_init` is a more fitting name.

File: src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.c [hunk 5]
```diff
@@ -33,6 +34,20 @@
...
+    void *mem = NULL;
+    /* Create the shared memory regions that will be used for the iqueue cells and terminals. */
+    MPIDU_Init_shm_alloc(size, &mem);
+    return mem;
+}
+
+void iqueue_free_fn(void *addr)
+{
+    MPIDU_Init_shm_free(addr);
+
```
[the remark this answers is not in the document]
Comment (yfguo) [reply src/mpid/ch4/shm/posix/eager/iqueue/iqueue_init.c:48]: Will clean this.

File: src/mpid/common/iqueue/mpidu_iqueue.h [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+#define MPIDU_IQUEUE_QUEUE_GET_BASE(queue) \
+    (void *) ((char *) (queue) - (queue)->u.handle)
+
+/* Grab a specific cell by using the offset value (from the beginning of the memory used to store
+ * all of the cells) of the handle. */
+#define MPIDU_IQUEUE_GET_CELL_BY_HANDLE(queue, handle) \
+    (MPIDU_iqueue_cell_t*)((handle) ? \
+        (MPIDU_IQUEUE_QUEUE_GET_BASE((queue)) + (uintptr_t)(handle)) : NULL)
+
+int MPIDU_iqueue_pool_alloc(int cell_size, int num_cells_in_block, int max_num_cellblocks,
```
[the remark this answers is not in the document]
Comment (yfguo) [reply src/mpid/common/iqueue/mpidu_iqueue.h:90]: Yes. The name is sort of strange right now, need to do a naming clean up through out the code. I was thinking about the case that we might need "queue blocks" in the future if we need to support multiple queues per process. In that case, the current static "get queue by index" scheme may need to be changed.

File: src/mpid/common/iqueue/mpidu_iqueue.c [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+    new_pool->cell_size = cell_size;
+    new_pool->num_cells_in_block = num_cells_in_block;
+    new_pool->max_num_cellblocks = max_num_cellblocks;
+    new_pool->max_num_queues = max_num_queues;
+
+    new_pool->malloc_fn = malloc_fn;
+    new_pool->free_fn = free_fn;
+    slab = malloc_fn(total_queue_size + sizeof(MPIDU_iqueue_pool_t)
+                     + total_queue_size + sizeof(MPIDU_iqueue_cellblock_mgr_t)
```
[the remark this answers is not in the document]
Comment (yfguo) [reply src/mpid/common/iqueue/mpidu_iqueue.c:45]: I messed up here. I am not putting the pool data structure on the slab. Will fix it.

File: src/mpid/common/iqueue/mpidu_iqueue.c (hunk 1 above)
Comment (yfguo) [reply src/mpid/common/iqueue/mpidu_iqueue.c:65]: 👍

File: src/mpid/common/iqueue/mpidu_iqueue.h [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+#define MPIDU_IQUEUE_SP     (0x100)
+#define MPIDU_IQUEUE_MP     (0x1000)
+
+typedef void *(*MPIDU_iqueue_malloc_fn) (uintptr_t);
+typedef void (*MPIDU_iqueue_free_fn) (void *);
+
+typedef struct MPIDU_iqueue_cell MPIDU_iqueue_cell_t;
+
+/* Each cell contains some data being communicated from one process to another. */
+struct MPIDU_iqueue_cell {
```
Comment (yfguo) [inline src/mpid/common/iqueue/mpidu_iqueue.h:34]: should be "cell_header"

**Comment by pavanbalaji** [comment]
It seems a bit weird to call both the SHM transport and the MPIDU framework as `iqueue`.  Can we rename one of them?  Since the term `iqueue` is already taken for the shm transport, how about we call the MPIDU infrastructure something else?

**Comment by yfguo** [comment]
Yeah. The name is confusing. maybe we could call it`MPIDU_queue`.

**Comment by pavanbalaji** [comment]
> Yeah. The name is confusing. maybe we could call it`MPIDU_queue`.

So the functions will be `MPIDU_queue_shared_queue_enqueue`?  The naming sounds a bit crazy.  :-)

**Comment by yfguo** [comment]
> > Yeah. The name is confusing. maybe we could call it`MPIDU_queue`.
> 
> So the functions will be `MPIDU_queue_shared_queue_enqueue`? The naming sounds a bit crazy. :-)

Couldn't come up with a better name. I think I will just use `MPIDU_queue`.

`MPIDU_iqueue_shared_queue_t` will become `MPIDU_queue_shared_t`.
`MPIDU_iqueue_shared_queue_enqueue` will become `MPIDU_queue_shared_enqueue`.
Though, `MPIDU_iqueue_shared_cell_pool_t` will become `MPIDU_queue_shared_cell_pool_t`

**Comment by pavanbalaji** [comment]
> Couldn't come up with a better name. I think I will just use `MPIDU_queue`.
> 
> `MPIDU_iqueue_shared_queue_t` will become `MPIDU_queue_shared_t`.
> `MPIDU_iqueue_shared_queue_enqueue` will become `MPIDU_queue_shared_enqueue`.
> Though, `MPIDU_iqueue_shared_cell_pool_t` will become `MPIDU_queue_shared_cell_pool_t`

How will you distinguish queues and pools then, if everything is called a queue?

**Comment by yfguo** [comment]
Or use `MPIDU_gqueue` to replace `MPIDU_iqueue`. `g` for general.

**Comment by pavanbalaji** [comment]
> Or use `MPIDU_gqueue` to replace `MPIDU_iqueue`. `g` for general.

How about `MPIDU_genq`?

**Comment by yfguo** [comment]
> > Or use `MPIDU_gqueue` to replace `MPIDU_iqueue`. `g` for general.
> 
> How about `MPIDU_genq`?

That sounds fine. `MPIDU_genq` it is.

**Comment by yfguo** [comment]
@pavanbalaji This PR is updated following our discussion yesterday.

**Comment by pavanbalaji** [comment]
This looks clean to me.  Not signing off yet, because there are more things pending in this PR.

test:mpich/ch4/most

**Comment by yfguo** [comment]
test:mpich/ch4/most

**Comment by yfguo** [comment]
test:mpich/ch4/most

**Comment by yfguo** [comment]
test:mpich/ch4/most

**Comment by yfguo** [comment]
test:mpich/ch4/most

File: src/mpid/common/genq/mpidu_genq_common.h [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+ *     See COPYRIGHT in top-level directory
+ */
+
+#ifndef MPIDU_GENQ_COMMON_H_INCLUDED
+#define MPIDU_GENQ_COMMON_H_INCLUDED
+
+#include <stdint.h>
+
+typedef void *(*MPIDU_genq_malloc_fn) (uintptr_t);
+typedef void (*MPIDU_genq_free_fn) (void *);
```
Comment (pavanbalaji) [inline src/mpid/common/genq/mpidu_genq_common.h:12]: Do we need these for this PR?

Comment (yfguo) [reply src/mpid/common/genq/mpidu_genq_common.h:12]: No, it is only used in private pool. I will remove it.

**Comment by pavanbalaji** [comment]
test:mpich/ch4/most

**Comment by yfguo** [comment]
Approved-by: Pavan Balaji <balaji@anl.gov>

**Pull request merged**